# mask-alignment

Method to hard mask genome-specific sites in an existing whole genome alignment.

Input:
- Whole genome alignment in Multiple Alignment Format (MAF). Note: I have only tested this code on MAFs produced using the [cactus](https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/README.md) function `cactus-hal2maf`
- Fasta file with (unaligned) replacement genomes containing desired masking

Output:
- Fasta alignment with all replacement genomes (some sites masked)

### considerations
- This is a work in progress
- Sequences of MAF alignment blocks with zero-length genomic intervals will always be filtered

#### dependencies
- bedtools (tested using version 2.29.2)
- R v4+
- R packages `dplyr`, `stringr`, `Biostrings`, and `GenomicRanges`
- [settings.config](https://github.com/JeffWeinell/mask-alignment/blob/main/settings.config) download and edit to specify settings for `mask-alignment.sh`
- R script [mask-alignment.R](https://github.com/JeffWeinell/mask-alignment/blob/main/bin/mask-alignment.R) (download but don't edit)

#### Usage

The main bash script to run is [mask-alignment.sh](https://github.com/JeffWeinell/mask-alignment/blob/main/bin/mask-alignment.sh)

```
# to run:
/path/to/mask-alignment.sh /path/to/settings.config
```

#### Config file

Download and edit [settings.config](https://github.com/JeffWeinell/mask-alignment/blob/main/settings.config) before running `mask-alignment.sh`

The unedited `settings.config` file looks like this:

```
# Required variables to set
ALN_MAF_PATH=                       # path to input MAF generated by 'cactus-hal2maf'
GENOMES_PATH=                       # path to input fasta containing a MASKED version of each UNALIGNED/NO-GAPS genome. These genomes must all be in a single file and sequence names (in fasta headers) must follow the USCS naming convention of '>genome.chromosome'
ALN_FAOUT_PATH=                     # where to save output fasta
MASK_ALIGNMENT_RSCRIPT=             # path to your copy of 'mask-alignment.R'

# Optionally define where to look for installed R packages
R_PACKAGES_DIR=                     # path to directory where your R packages are installed (R default location used if left blank)

```








